frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1514 @@
+			self.dyn_child_items.saturating_add(self.dyn_top_items)
+		}
+
+		/// Migrate keys until either of the given limits are exhausted, or if no more top keys
+		/// exist.
+		///
+		/// Note that this returns after the **first** migration tick that causes exhaustion. In
+		/// other words, this should not be used in any environment where resources are strictly
+		/// bounded (e.g. a parachain), but it is acceptable otherwise (relay chain, offchain
+		/// workers).
+		pub(crate) fn migrate_until_exhaustion(&mut self, limits: MigrationLimits) {
+			log!(debug, "running migrations on top of {:?} until {:?}", self, limits);
+
+			if limits.item.is_zero() || limits.size.is_zero() {
+				// handle this minor edge case, else we would call `migrate_tick` at least once.
+				log!(warn, "limits are zero. stopping");
+				return
+			}
+
+			loop {
+				self.migrate_tick();
+				if self.exhausted(limits) {
+					break
+				}
+			}
+
+			// accumulate dynamic data into the storage items.
+			self.size = self.size.saturating_add(self.dyn_size);
+			self.child_items = self.child_items.saturating_add(self.dyn_child_items);
+			self.top_items = self.top_items.saturating_add(self.dyn_top_items);
+			log!(debug, "finished with {:?}", self);
+		}
+
+		/// Migrate AT MOST ONE KEY. This can be either a top or a child key.
+		///
+		/// This function is the core of this entire pallet.
+		fn migrate_tick(&mut self) {
+			match (self.last_top.as_ref(), self.last_child.as_ref()) {
+				(Some(_), Some(_)) => {
+					// we're in the middle of doing work on a child tree.
+					self.migrate_child();
+				},
+				(Some(ref top_key), None) => {
+					// we have a top key and no child key. 3 possibilities exist:
+					// 1. we continue the top key migrations.
+					// 2. this is the root of a child key, and we start processing child keys (and
+					// should call `migrate_child`).
+					// 3. this is the root of a child key, and we are finishing all child-keys (and
+					// should call `migrate_top`).
+
+					// NOTE: this block is written intentionally to verbosely for easy of
+					// verification.
+					match (
+						top_key.starts_with(DEFAULT_CHILD_STORAGE_KEY_PREFIX),
+						self.prev_tick_child,
+					) {
+						(false, false) => {
+							// continue the top key migration
+							self.migrate_top();
+						},
+						(true, false) => {
+							// start going into a child key.
+							let maybe_first_child_key = {
+								let child_top_key = Pallet::<T>::child_io_key(top_key);
+								sp_io::default_child_storage::next_key(child_top_key, &vec![])
+							};
+							if let Some(first_child_key) = maybe_first_child_key {
+								self.last_child = Some(first_child_key);
+								self.prev_tick_child = true;
+								self.migrate_child();
+							} else {
+								self.migrate_top();
+							}
+						},
+						(true, true) => {
+							// we're done with migrating a child-root.
+							self.prev_tick_child = false;
+							self.migrate_top();
```

Comment:
note that this will migrate the top key which store the root of the child trie, writing on such a key is forbidden and will result in no-op. But it will trigger a warning https://github.com/paritytech/substrate/blob/07d98e28cd7a79e45f02a982af14fbde929d8a16/primitives/state-machine/src/ext.rs#L391
maybe it is better to not trigger warnings, but it can be fine.

Reply:
Calling sp_io::storage::set only for value of length at least 33 bytes, also make this unreachable.

Reply:
indeed we should skip the write for value less than threshold

Reply:
I've run migrations on the entire polkadot/kusama state (both have child trie now) and I've never seen this warning. Any explanation?